c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -15,25 +20,123 @@ mod dump;
 mod type_check;
 
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Default)]
-pub struct Label {
+pub struct Label<'tcx> {
+    /// The [`Origin`] of this type
     pub origin: Option<Origin>,
+    /// The [Origins](`Origin`) associated with each lifetime
+    /// parameter of this type, if applicable
+    pub origin_params: Option<&'tcx [(OriginParam, Origin)]>,
     pub perm: PermissionSet,
 }
 
-pub type LTy<'tcx> = LabeledTy<'tcx, Label>;
-pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label>;
+pub type LTy<'tcx> = LabeledTy<'tcx, Label<'tcx>>;
+pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label<'tcx>>;
+
+/// Metadata describing lifetimes and lifetime parameters
+/// of a struct field.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FieldMetadata<'tcx> {
+    /// The lifetime parameters of a field, e.g. if a struct
+    /// `foo<'a, 'b>` is a field of `bar<'c, 'd>` as field: `foo<'c, 'd>`,
+    /// the lifetime params would be a set {'c, 'd}. For a reference such
+    /// as &'r foo<'c, 'd>, the lifetime params in the label would be
+    /// {'r}, and {'c, 'd} would be the label of the sole argument
+    /// of the labeled reference type
+    pub lifetime_params: LabeledTy<'tcx, &'tcx [OriginArg<'tcx>]>,
+}
+
+/// Metadata describing the lifetime parameters and fields
+/// of a struct.
+#[derive(Clone, PartialEq, Eq)]
+pub struct AdtMetadata<'tcx> {
```

Comment:
Is this for `struct`s only?  Why call it `AdtMetadata` then?  Also, in general, how does this PR interact with non-`struct` ADTs (`enum`s)?

Reply:
for now, it's tailored to `structs` and `unions`. `enums` are not yet supported in full. essentially, anything that can be coerced via `non_enum_variant` is supported and will otherwise panic